### PR TITLE
Create new alphabet class: Enumerated alphabet

### LINKED ISSUE #199: Create enumerated alphabet class
As agreed upon, we should implement an enumerated alphabet class which acts as the current `IntAlphabet`, but only enumerates the requested symbols. This enumerated alphabet does not implicitly contain all integer symbols.

Furthermore, the default implicitly created alphabet should be unset (`Nfa::alphabet` is set to `nullptr`) from all `Nfa` constructors and constructor functions unless a specific alphabet is provided by the user or the function truly expects a specific alphabet to be created.

### PR DESCRIPTION
**Pull request by Adda0**
This PR implements an enumerated alphabet class, `Mata::EnumAlphabet`, which holds a set of symbols, behaving like `Mata::IntAlphabet`, only working with a limited set of symbols instead of presumably the whole `Mata::Symbol` value range.

In order for the enumerated alphabet class (and its methods) to work, this PR implements some support functions and fixes inconsistencies in `Mata::Util::OrdVector` which `Mata:EnumAlphabet` depends on.

Resolves #199.

### CODE REVIEW

File: include/mata/alphabet.hh [hunk 1]
```diff
@@ -169,10 +169,151 @@ namespace Mata {
...
+            }
+            return std::to_string(symbol);
+        }
+
+    private:
+        EnumAlphabet& operator=(const EnumAlphabet& rhs);
+
+    public:
+        /**
+         * @brief Expand alphabet by symbols from the passed @p symbol_names.
```
Comment (jurajsic) [inline include/mata/alphabet.hh]:
```suggestion
         * @brief Expand alphabet by symbols from the passed @p symbols.
```

File: include/mata/alphabet.hh [hunk 2]
```diff
@@ -169,10 +169,151 @@ namespace Mata {
...
+        EnumAlphabet(std::initializer_list<std::string> l) : EnumAlphabet(l.begin(), l.end()) {}
+
+        Symbol translate_symb(const std::string& str) override {
+            Symbol symbol;
+            std::istringstream stream{ str };
+            stream >> symbol;
+            return symbol;
+
+            // TODO: How can the user specify to throw exceptions when we encounter an unknown symbol? How to specify that
+            //  the alphabet should have only the previously fixed symbols?
```
Comment (jurajsic) [inline include/mata/alphabet.hh]: I think we should just throw an exception if we are trying to translate "unknown" symbol.

**Comment by Adda0** [comment]
`OrdVector::Resize()` is replaced with `OrdVector::reserve()`, because resizing ordered container does not make any sense. Resize to smaller number would remove the last largest `size - n` elements. Resizing to a larger number would default-insert values, zeros in our case, which would break sortedness and invariant of only one occurrence in the set. This is a nasty hidden "feature" which is error-prone in case of sorted sets.

What we want from this method anyway is to preallocate the memory for some known number of elements in advance so that we do not have to reallocate the vector multiple times. `OrdVector::reserve()` does exactly that while keeping the actual content of the vector unmodified.

**Review by jurajsic — APPROVED** [review]
Looks ok.

File: include/mata/alphabet.hh (hunk 2 above)
Comment (Adda0) [reply include/mata/alphabet.hh]: I agree. I will modify it and if no errors arise, I will commit the changes.

File: include/mata/alphabet.hh (hunk 1 above)
Comment (Adda0) [reply include/mata/alphabet.hh]: Will do.

File: include/mata/alphabet.hh (hunk 2 above)
Comment (Adda0) [reply include/mata/alphabet.hh]: Modified.

File: include/mata/alphabet.hh [hunk 3]
```diff
@@ -169,10 +172,155 @@ namespace Mata {
             IntAlphabetSingleton &alphabet_instance;
         }; // class IntAlphabet.
 
-/**
- * An alphabet constructed 'on the fly'.
- * Should be use anytime the automata have a specific names for the symbols.
- */
+    /**
+     * Enumerated alphabet using a set of integers as symbols maintaining a set of specified symbols.
```
Comment (tfiedor) [inline include/mata/alphabet.hh:176]: I'd suggest in the comment some example of some EnumAlphabet...like, e.g. `Example of alphabet = {"1", "2", "3", ... }`.

**Review by tfiedor — APPROVED** [review]
No issues

File: include/mata/alphabet.hh (hunk 3 above)
Comment (Adda0) [reply include/mata/alphabet.hh:176]: Great idea. Will do.

**Comment by Adda0** [comment]
PR is rebased onto the current `devel`. Conflicts with `devel` are resolved. The PR is ready to be merged.

**Pull request merged**